src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs

```diff
@@ -545,22 +537,45 @@ private void Poll()
             }
         }
 
-        private List<ConsumeResult<K, V>> PollKafka(CancellationToken token)
+        private void ProcessExceptions(Exception exception)
+        {
+            if (exception == null)
+                return;
+
+            var directive = _decider(exception);
+            ProcessError(exception);
+            if (directive == Directive.Resume)
+                return;
+            
+            _pollCancellation?.Cancel();
+            if(directive == Directive.Stop && _log.IsErrorEnabled)
+                _log.Error(exception, "Exception when polling from consumer, stopping actor: {0}", exception.Message);
+            Context.Stop(Self);
+        }
+
+        private (List<ConsumeResult<K, V>>, Exception) PollKafka(CancellationToken token)
         {
             ConsumeResult<K, V> consumed = null;
             var i = 10; // 10 poll attempts
             var timeout = Math.Max((int) _pollTimeout.TotalMilliseconds / i, 1);
-            var pooled = new List<ConsumeResult<K, V>>();
+            var polled = new List<ConsumeResult<K, V>>();
             do
             {
-                // this would return immediately if there are messages waiting inside the client queue buffer
-                consumed = _consumer.Consume(timeout);
+                try
+                {
+                    // this would return immediately if there are messages waiting inside the client queue buffer
+                    consumed = _consumer.Consume(timeout);
+                }
+                catch (Exception e)
+                {
+                    return (polled, e);
+                }
                 if (consumed != null)
-                    pooled.Add(consumed);
+                    polled.Add(consumed);
                 i--;
-            } while (i > 0 && !token.IsCancellationRequested);
+            } while (i > 0 && consumed != null && !token.IsCancellationRequested);
```

Comment:
So this will exit right away as soon as we receive a single message - no point in building a list of them if that's the case, no?

Reply:
No, Consumer.Consume will return a null if its internal message queue is empty. So instead of waiting for the whole timeout duration, we just bail out of the polling method and return the messages that we polled up to that point.

Reply:
You can consider the Consumer.Consume returning null the same as C string null terminator

Reply:
Ok got it - I read the boolean logic backwards there